### PR TITLE
Added note about using the Validator TypeTestCase in form unit testing

### PR DESCRIPTION
**Pull request by pierredup**
This update ads the usage of the validator TypeTestCase class to use for testing when validation is required.

### CODE REVIEW

**Review by HeahDude — APPROVED** [review]
Thanks!

**Comment by xabbuh** [comment]
Actually, after thinking about this again I am 👎 on merging the PR as is. We usually don't want to guarantee BC for classes that live inside the `Tests` namespace (that's why there is a `Test` namespace that contains classes that are meant to be used as base classes for userland code).

**Comment by pierredup** [comment]
What would you suggest needs to change? Should the clas be moved to the `Test` namespace under a different name (Something like `ValidatorTestCase`)?

**Comment by xabbuh** [comment]
I am not completely settled on that yet. The main issue is: What if you need another extension? Which base class do you extended then (supposed there is a base class for the other extension as well)?

**Comment by xabbuh** [comment]
I think the best is to open an issue on the code repository and discuss the matter over there first.

**Comment by wouterj** [comment]
While I like your suggestion, I don't think we should document classes that live in `*\Tests\*`. These classes aren't part of the BC promise and thus shouldn't be used by other packages.

Instead, we should update our example to fix the code. This probably means finishing https://github.com/symfony/symfony-docs/pull/6463 .

However, if it turns out to be a very nice addition to unit testing, the class might need to be moved to `Symfony\Component\Form\Test\ValidatorTypeTestCase` so it can be used by other packages. As I've seen this class used by quite a few packages, I suggest submitting a PR doing so against the code repository. Are you maybe willing to do that?

**Comment by pierredup** [comment]
Hence my original PR that added the validator to the main base class with an extension point to register custom extensions (and other core extensions can be added in the same place).
I'll open another PR following that approach again and see what discussions come out of it

**Comment by wouterj** [comment]
Sorry if I duplicated something that's already done. I lost overview with all these referenced PRs. :innocent:

**Comment by pierredup** [comment]
I would suggest updating the standard base class to include the validator extension, and remove (deprecate?) the one under `tests` so that there is one class to extend, both for userland and internal tests (also checking that the validator is installed before registering the extension).
I'll open a PR on the code repository for this and wait for feedback

**Comment by wouterj** [comment]
> I would suggest updating the standard base class to include the validator extension

I don't think this will be accepted by the core team. The form component can be used standalone without the validator, so it should at least be optional. You may create a trait with a `getValidatorExtension()` method that can be easily called when extending the `TypeTestCase`?

**Comment by pierredup** [comment]
> also checking that the validator is installed before registering the extension

This should cover the case where the form component is used in standalone mode and the validator is not installed.

File: form/unit_testing.rst
Comment (xabbuh) [inline form/unit_testing.rst]: @pierredup Can you remove this section? We can then still merge your fixes to the class namespaces which is still a legit fix.

File: form/unit_testing.rst
Comment (pierredup) [inline form/unit_testing.rst]: Should these changes also be reverted?

Comment (xabbuh) [inline form/unit_testing.rst]: I think so

**Comment by javiereguiluz** [comment]
What can we do to make this PR mergeable? Thanks!

**Comment by HeahDude** [comment]
@javiereguiluz Revert everything excepted namespaces modifications like https://github.com/symfony/symfony-docs/pull/7587/files#diff-d9a34263773b4fe7719e18118a10e628R204.

**Comment by javiereguiluz** [comment]
@pierredup we had to revert part of your proposed changes, but merged others. Thanks for your contribution!

**Pull request merged**
